sdk/keyvault/azsecrets/client.go

```diff
@@ -0,0 +1,1011 @@
+	PerTryPolicies []policy.Policy
+}
+
+func (c *ClientOptions) toConnectionOptions() *internal.ConnectionOptions {
+	if c == nil {
+		return nil
+	}
+
+	return &internal.ConnectionOptions{
+		HTTPClient:       c.HTTPClient,
+		Retry:            c.Retry,
+		Telemetry:        c.Telemetry,
+		Logging:          c.Logging,
+		PerCallPolicies:  c.PerCallPolicies,
+		PerRetryPolicies: c.PerTryPolicies,
+	}
+}
+
+// NewClient returns a pointer to a Client object affinitized to a vaultUrl.
+func NewClient(vaultUrl string, credential azcore.TokenCredential, options *ClientOptions) (*Client, error) {
+	if options == nil {
+		options = &ClientOptions{}
+	}
+
+	conn := internal.NewConnection(credential, options.toConnectionOptions())
+
+	return &Client{
+		kvClient: &internal.KeyVaultClient{
+			Con: conn,
+		},
+		vaultUrl: vaultUrl,
+	}, nil
+}
+
+// GetSecretOptions holds the optional parameters for the Client.GetSecret function
+type GetSecretOptions struct {
+	// Version specifies the version of a secret. If unspecified, the most recent version will be returned
+	Version string
+}
+
+// convert the exposed options struct to the internal one.
+func (g *GetSecretOptions) toGenerated() *internal.KeyVaultClientGetSecretOptions {
+	if g == nil {
+		return &internal.KeyVaultClientGetSecretOptions{}
+	}
+	return &internal.KeyVaultClientGetSecretOptions{}
+}
+
+// GetSecretResponse is the response object for the Client.GetSecret operation
+type GetSecretResponse struct {
+	SecretBundle
+
+	// RawResponse contains the underlying HTTP response
+	RawResponse *http.Response
+}
+
+func getSecretResponseFromGenerated(i internal.KeyVaultClientGetSecretResponse) *GetSecretResponse {
+	return &GetSecretResponse{
+		RawResponse: i.RawResponse,
+		SecretBundle: SecretBundle{
+			Attributes:  secretAttributesFromGenerated(i.Attributes),
+			ContentType: i.ContentType,
+			ID:          i.ID,
+			Tags:        i.Tags,
+			Value:       i.Value,
+			Kid:         i.Kid,
+			Managed:     i.Managed,
+		},
+	}
+}
+
+// GetSecret gets a specified secret from a given key vault. The GET operation is applicable to any secret
+// stored in Azure Key Vault. This operation requires the secrets/get permission
+func (c *Client) GetSecret(ctx context.Context, secretName string, options *GetSecretOptions) (GetSecretResponse, error) {
+	if options == nil {
+		options = &GetSecretOptions{}
+	}
+	resp, err := c.kvClient.GetSecret(ctx, c.vaultUrl, secretName, options.Version, options.toGenerated())
+	return *getSecretResponseFromGenerated(resp), err
+}
+
+// SetSecretOptions contains the optional parameters for a Client.SetSecret operation
+type SetSecretOptions struct {
+	// Type of the secret value such as a password.
+	ContentType *string `json:"contentType,omitempty"`
+
+	// The secret management attributes.
+	SecretAttributes *SecretAttributes `json:"attributes,omitempty"`
+
+	// Application specific metadata in the form of key-value pairs.
+	Tags map[string]*string `json:"tags,omitempty"`
```

Comment:
Do we really want *string for the value type? Can't this be just string

Reply:
this is a copy of what the generated code does, I can make this change in the public API I believe but it would add extra cycles

Reply:
I think "Tags *map[string]string" (* before map; not after ]) supports our nullable common denominator.
Setting Tags to null is how we'd delete the tags field for JSON patch (which is not supported for this particular scenario).
In this case, setting the value to null has no meaning which is why this looks wrong to me.
I'm also concerned about the round tripping: it might be useful to get Tags for 1 blob and then apply them to a new blob and so both the GET & PUT need a Tags map of the same type (map[string]string).

Reply:
The pattern we have today is that to remove an item you would specify `fooMap["key"] = nil` which sends `{"fooMap": { "key": null } }` across the wire.  Doesn't tags behave this way?

Reply:
Yes, I see your point. No, Tags do not work like this. The Storage service doesn't support PATCH at all (and probably never will).
We could leave this as is as long as the round-tripping works easily. But, I do think the *string hurts usability for Tags since setting to null is not a thing.

Reply:
I see.  Agreed that it hurts the usability then and should be removed in the public API.